homeassistant/components/icloud/config_flow.py

```diff
@@ -0,0 +1,243 @@
+    DEFAULT_GPS_ACCURACY_THRESHOLD,
+    DOMAIN,
+)
+
+CONF_TRUSTED_DEVICE = "trusted_device"
+CONF_VERIFICATION_CODE = "verification_code"
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class IcloudFlowHandler(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a iCloud config flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_CLOUD_POLL
+
+    def __init__(self):
+        """Initialize iCloud config flow."""
+        self.api = None
+        self._username = None
+        self._password = None
+        self._accountname = None
+        self._max_interval = None
+        self._gps_accuracy_threshold = None
+
+        self._trusted_device = None
+        self._verification_code = None
+
+    def _configuration_exists(self, username: str, accountname: str) -> bool:
+        """Return True if username or accountname exists in configuration."""
+        for entry in self.hass.config_entries.async_entries(DOMAIN):
+            if (
+                entry.data[CONF_USERNAME] == username
+                or entry.data[CONF_ACCOUNT_NAME] == accountname
+                or entry.data[CONF_ACCOUNT_NAME] == slugify(username.partition("@")[0])
+            ):
+                return True
+        return False
+
+    async def _show_setup_form(self, user_input=None, errors=None):
+        """Show the setup form to the user."""
+
+        if user_input is None:
+            user_input = {}
+
+        return self.async_show_form(
+            step_id="user",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(
+                        CONF_USERNAME, default=user_input.get(CONF_USERNAME, "")
+                    ): str,
+                    vol.Required(
+                        CONF_PASSWORD, default=user_input.get(CONF_PASSWORD, "")
+                    ): str,
+                }
+            ),
+            errors=errors or {},
+        )
+
+    async def async_step_user(self, user_input=None):
+        """Handle a flow initiated by the user."""
+        errors = {}
+
+        icloud_dir = self.hass.config.path("icloud")
+        if not os.path.exists(icloud_dir):
+            self.hass.async_add_executor_job(os.makedirs, icloud_dir)
+
+        if user_input is None:
+            return await self._show_setup_form(user_input, errors)
+
+        self._username = user_input[CONF_USERNAME]
+        self._password = user_input[CONF_PASSWORD]
+        self._accountname = user_input.get(CONF_ACCOUNT_NAME)
+        self._max_interval = user_input.get(CONF_MAX_INTERVAL, DEFAULT_MAX_INTERVAL)
+        self._gps_accuracy_threshold = user_input.get(
+            CONF_GPS_ACCURACY_THRESHOLD, DEFAULT_GPS_ACCURACY_THRESHOLD
+        )
+
+        if self._configuration_exists(self._username, self._accountname):
+            errors[CONF_USERNAME] = "username_exists"
+            return await self._show_setup_form(user_input, errors)
+
+        try:
+            self.api = PyiCloudService(
```

Comment:
What api calls does I/O? All the steps in the config flow are run in the event loop, so must not block.

Reply:
It is doing HTTP call yes, but I should do it here !

How can I do smarter ?

Reply:
Use the executor, `hass.async_add_executor_job`.